packages/parser/src/BattlegroundData.ts

```diff
@@ -0,0 +1,414 @@
+    return cls.abilities.map((ability) => [ability.spellId, cls.unitClass]);
+  }),
+);
+
+export class BattlegroundData {
+  public id: string = uuidv4();
+  public isWellFormed = false;
+  public startTime = 0;
+  public endTime = 0;
+  public units: { [unitId: string]: CombatUnit } = {};
+  public playerId = '';
+  public playerTeamId = '';
+  public hasAdvancedLogging = false;
+  public rawLines: string[] = [];
+  public linesNotParsedCount = 0;
+  public events: (CombatAction | CombatAdvancedAction)[] = [];
+
+  private combatantMetadata: Map<string, ICombatantMetadata> = new Map<string, ICombatantMetadata>();
+
+  private inferredCombatantIds: Set<string> = new Set<string>();
+
+  constructor(
+    public readonly wowVersion: WowVersion,
+    public readonly timezone: string,
+  ) {}
+
+  public readEvent(event: CombatEvent) {
+    if (this.startTime === 0) {
+      this.startTime = event.timestamp;
+    }
+    this.endTime = event.timestamp;
+
+    if (event instanceof ZoneChange) {
+      return;
+    }
+
+    if (event.logLine.parameters.length < 8) {
+      return;
+    }
+
+    if (event instanceof CombatAction) {
+      this.events.push(event);
+    }
+
+    if (event instanceof ArenaMatchEnd) {
+      throw new Error('This is not possible but needed for TS');
+    }
+    if (event instanceof ArenaMatchStart) {
+      throw new Error('This is not possible but needed for TS');
+    }
+    if (event instanceof CombatantInfoAction) {
+      throw new Error('This is not possible but needed for TS');
+    }
+
+    const srcGUID = event.srcUnitId;
+    const srcName = event.srcUnitName;
+    const srcFlag = event.srcUnitFlags;
+
+    const destGUID = event.destUnitId;
+    const destName = event.destUnitName;
+    const destFlag = event.destUnitFlags;
+
+    if (!this.units[srcGUID]) {
+      this.units[srcGUID] = new CombatUnit(srcGUID, srcName);
+    }
+    if (!this.units[destGUID]) {
+      this.units[destGUID] = new CombatUnit(destGUID, destName);
+    }
+
+    const srcUnit = this.units[srcGUID];
+    const destUnit = this.units[destGUID];
+    if (!srcUnit || !destUnit) {
+      throw new Error('failed to parse source unit or dest unit from the log line');
+    }
+    srcUnit.endTime = event.timestamp;
+    destUnit.endTime = event.timestamp;
+
+    srcUnit.proveType(getUnitType(srcFlag));
+    destUnit.proveType(getUnitType(destFlag));
+
+    srcUnit.proveReaction(getUnitReaction(srcFlag));
+    destUnit.proveReaction(getUnitReaction(destFlag));
+
+    const isSignificantEvent =
+      event.srcUnitId !== event.destUnitId &&
+      getUnitReaction(event.srcUnitFlags) !== getUnitReaction(event.destUnitFlags) &&
+      getUnitType(event.srcUnitFlags) === CombatUnitType.Player &&
+      getUnitType(event.destUnitFlags) === CombatUnitType.Player;
+
+    if (isSignificantEvent) {
+      this.inferredCombatantIds.add(event.srcUnitId);
+      this.inferredCombatantIds.add(event.destUnitId);
+    }
+
+    switch (event.logLine.event) {
```

Comment:
a lot of the logic should be consistent with the other combat class. might be good to extract to a base class and reuse?

Reply:
yeah I need to have a think through this. there's definitely going to be some common code but also a lot of code that *looks* common but has super subtle diffs